nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -208,19 +239,34 @@ void KeepoutFilter::process(
 
   unsigned int i, j;  // master_grid iterators
   unsigned int index;  // corresponding index of master_grid
+  double gl_wx, gl_wy;  // world coordinates in a global_frame_
+  double msk_wx, msk_wy;  // world coordinates in a mask_frame_
   unsigned int mx, my;  // mask_costmap_ coordinates
   unsigned char data, old_data;  // master_grid element data
 
   // Main master_grid updating loop
-  // Iterate in overlapped window by master_grid indexes
+  // Iterate in costmap window by master_grid indexes
   unsigned char * master_array = master_grid.getCharMap();
   for (i = mg_min_x_u; i < mg_max_x_u; i++) {
     for (j = mg_min_y_u; j < mg_max_y_u; j++) {
       index = master_grid.getIndex(i, j);
       old_data = master_array[index];
-      // Calculating corresponding to (i, j) point at mask_costmap_
-      master_grid.mapToWorld(i, j, wx, wy);
-      if (mask_costmap_->worldToMap(wx, wy, mx, my)) {
+      // Calculating corresponding to (i, j) point at mask_costmap_:
+      // Get world coordinates in global_frame_
+      master_grid.mapToWorld(i, j, gl_wx, gl_wy);
+      if (mask_frame_ != global_frame_) {
+        // Transform (i, j) point from global_frame_ to mask_frame_
+        tf2::Vector3 point(gl_wx, gl_wy, 0);
```

Comment:
Would it not be more adventitious to convert to the new frame on reception of a costmap rather than owning this compute every update cycle?

Reply:
Then, none of the changes in the `process` function would be necessary, just in the costmap callback we check if frames are different, convert the mask to the costmap's global frame, and all else continues as normal as if it originally came in in that frame.

Reply:
Mhm, though this wouldn't allow for moving frames....  is that something we think we'd need to / want to support?

Reply:
> Mhm, though this wouldn't allow for moving frames.... is that something we think we'd need to / want to support?

Unfortunately, yes. The simple example I've faced with - is that mask is being published in "map" frame whether the costmap is in "odom". "odom"->"map" transform is being managed by AMCL, and this TF is changing dynamically in a time. So in general, at each `process()` call we need to re-calculate our points. You can find something similar in `static_layer.cpp` in `StaticLayer::updateCosts()` routine whether depending on `isRolling()` attribute (practically, different frames when rolling) each point from `(min_i, max_j)` window is being translated into map frame space.

Reply:
So you're seeing the keepout zones in the local costmap essentially drift over time? That's not ideal :laughing: